### ISSUE TITLE
Memory consumption in angr for our use cases is quite bad in newer versions

### ISSUE THREAD

**Issue by TodAmon** [angr/angr#4418]
### Description

We use angr to perform fairly large studies. Some of those studies can no longer be formed in the newer versions of angr.  The primary culprit is memory usage.  For our studies, new versions of angr run out of memory (consume 32GB plus) whereas old versions never used even a gigabyte.  I was able to create an example that demonstrates the problem.  Here are some data:
```
9.2.83 angr length 10 peak 290 MB leaked 132 MB over 03 min 34 sec
9.2.83 angr length 12 peak 495 MB leaked 236 MB over 06 min 31 sec
9.2.83 angr length 14 peak 994 MB leaked 446 MB over 10 min 59 sec
9.2.83 angr length 16 peak 1.2 GB leaked 537 MB over 17 min 10 sec

9.0.10730 angr length 10 peak 142 MB leaked 64 MB over 1 min 24 sec
9.0.10730 angr length 12 peak 147 MB leaked 68 MB over 1 min 39 sec
9.0.10730 angr length 14 peak 151 MB leaked 71 MB over 1 min 56 sec
9.0.10730 angr length 16 peak 200 MB leaked 108 MB over 2 min 15 sec
```
The "leaked" memory might just be consumed memory.  I used "heaptrack" to obtain this data.  In every case, it reports the leaked memory as primarily arising from libz3.so allocate and reallocate -- but of course it may just be that it wasn't garbage collected at termination.  So, there might not be a memory leak.  However, these numbers correspond well with "top" -- e.g., the new version of angr really is using vast amounts of memory compared to previous versions, and even worse as the complexity of the example goes up, the memory usage goes up at even worse rate.  For smaller examples, newer versions of angr often and usually out perform the old version.  But not for this example.  

### Steps to reproduce the bug

Simply run the angr script on the attached binary, passing "-l X" where X is symbolic string length (e.g. 10 - 16) in my tests
[check_char.tar.gz](https://github.com/angr/angr/files/13923830/check_char.tar.gz)

### Additional context

I previously discussed this issue with @ltfish.  Hopefully you can reproduce.  I tried to create a smaller example, but this was the best I could come up with.  The issue seems to require state merging and sim procedures, to rear its ugly head.

**Comment by ltfish** [comment]
Hi @TodAmon , I've been looking into this issue, and I have some interesting findings so far.

- By using angr (and all its dependencies) 9.0.10730 from pip with Ubuntu 20.04, Python 3.8, and z3 4.8.5.0, the peak memory usage (RES in htop) is 1.4 GB for `-l 16`. It took 8 minutes to finish.
- By using the latest angr (and all its dependencies) on GitHub with Python 3.11 and z3 4.10.2, the peak memory usage is 2.3 GB for `-l 16`. It took four minutes to finish.

I understand that Python 3.11 is way faster than 3.8, but the difference should not be so big when the majority of time of this test is spent in constraint solving inside z3. Anyway, it seems that the latest angr runs much faster for me than angr 9.0.10730, and I do not think I can reproduce your test result on my local machine.

Granted, the latest angr is using more RAM than the old angr. I can go ahead and attempt to reduce the RAM usage, but I also want to see if you can provide any insights regarding why angr 9.2.83 took so long and angr 9.0.10730 took so little time during your tests.

**Comment by ltfish** [comment]
I tried different versions of Z3 and measured their run time and RAM usage (all for `-l 16`, Python 3.11, latest angr on master, and Windows 11):

4.12.5.0: 48 seconds, 338 MB
4.12.0.0: 55 seconds, 400 MB
4.11.2.0: 284 seconds, 2.4 GB
4.11.0.0: 223 seconds, 2.5 GB
4.9.0.0: 213 seconds, 2.3 GB
4.8.10.0: 243 seconds, 2.6 GB
4.8.5.0: 152 seconds, 1.4 GB

I can't use Z3 before 4.8.5.0 because modern angr does not support Z3 that is older than 4.8.5.0.

So honestly it looks like a regression in Z3. Can you switch to the latest Z3 and try again?

**Comment by ltfish** [comment]
One more data point:

Python 3.8, angr 9.0.10730, Z3 4.12.5.0, and Ubuntu 20.04: 181 seconds and 446 MB of peak RAM usage.

So I would use the latest CPython + the latest angr + the latest Z3, which seem to yield the best performance. Can you please test this setup in your environment?

**Comment by ltfish** [comment]
There are too many commits in Z3 4.12.0.0 over Z3 4.11.2.0, and I am not interested in bisecting Z3 to figure out what the culprit was. I suspect that https://github.com/Z3Prover/z3/commit/d415f073866b4dc729587d3ce8fe3e2166f7ac02 fixes the memory leak.

**Comment by TodAmon** [comment]
Hi @ltfish thanks for the analysis.  I installed python 3.12, and Z3 4.12.5.0 but doing so uninstall claripy.  Installing claripy sends me back to z3 4.10.2.0 which is pinned in [https://github.com/angr/claripy/blob/master/setup.cfg](https://github.com/angr/claripy/blob/master/setup.cfg) I must be missing something, how did you try all those different versions?

**Comment by TodAmon** [comment]
Python 3.11 angr 9.2.86, z3: 4.10.2.0: 227 seconds, 2.4GB
Python 3.11 angr 9.2.86, z3: 4.12.5.0, 44 seconds,  352MB

So, for Python3.11, I am able to replicate some of your results.  More investigation to follow.

**Comment by ltfish** [comment]
I believe claripy fails in some corner cases for z3 >= 4.11.0.0. We can investigate bringing claripy up-to-date with Z3. I simply manually installed z3 by doing `pip install z3-solver==<z3_version_number>` in my angr virtualenv.

**Comment by TodAmon** [comment]
Yeah, that's what I did too, but it didn't work for python 3.12, worked for python 3.11, was my experience.  I tried the latest on our actual problem (e.g., not the toy above) and alas it made almost no additional progress before running out of memory.  Still, it does seem like some sort of progress.  I'm still confused about how the same version of z3, but different versions of angr, can have such different results... I need to see if I can reproduce the results I quoted on different machines, etc.  So, lets keep this open as I investigate more.

**Comment by ltfish** [comment]
@TodAmon Can you please see if https://github.com/angr/angr/pull/4448 fixes your issue? If it does, I suspect that you have been writing large consecutive chunks of data to the `SimConcreteFilesystem` in your angr project.

**Comment by TodAmon** [comment]
I tried angr 9.2.89 and our use cases are still blowing up with regards to memory usage.
